In [1]:
import numpy as np
from helpers import *
from collections import defaultdict
from implementations import *
import os

In [2]:
#Load data
x_train, x_test, y_train, train_ids, test_ids = load_csv_data("dataset")

## 1. Inspect raw data

In [3]:
print("x_train shape:", x_train.shape)
print("x_test shape:", x_test.shape)
print("y_train shape:", y_train.shape)
print("y_train unique values:", np.unique(y_train))

# dtype
print("x_train dtype:", x_train.dtype)

# NaN count per column (if NaNs are already np.nan)
nan_counts = np.isnan(x_train).sum(axis=0)
print("Columns with NaNs:", np.sum(nan_counts > 0), "out of", x_train.shape[1])
print("Max NaN fraction in a column:", nan_counts.max() / x_train.shape[0])

# Basic stats per column
print("Min per column (first 10):", np.nanmin(x_train[:, :10], axis=0))
print("Max per column (first 10):", np.nanmax(x_train[:, :10], axis=0))

x_train shape: (328135, 321)
x_test shape: (109379, 321)
y_train shape: (328135,)
y_train unique values: [-1  1]
x_train dtype: float64
Columns with NaNs: 239 out of 321
Max NaN fraction in a column: 0.9999024791625398
Min per column (first 10): [1.000000e+00 1.000000e+00 1.012016e+06 1.000000e+00 1.000000e+00
 2.015000e+03 1.100000e+03 2.015000e+09 2.015000e+09 1.000000e+00]
Max per column (first 10): [7.20000000e+01 1.20000000e+01 1.23120150e+07 1.20000000e+01
 3.10000000e+01 2.01600000e+03 1.20000000e+03 2.01502324e+09
 2.01502324e+09 1.00000000e+00]


## 2. Data cleaning
### 2.1 Distribution of NaN fractions across columns

In [4]:
nan_frac = np.isnan(x_train).sum(axis=0) / x_train.shape[0]

# Histogram-style summary of how "bad" columns are
print("Columns with 0% NaN:", np.sum(nan_frac == 0))
print("Columns with <5% NaN:", np.sum((nan_frac > 0) & (nan_frac < 0.05)))
print("Columns with 5-50% NaN:", np.sum((nan_frac >= 0.05) & (nan_frac < 0.5))) 
print("Columns with 50-90% NaN:", np.sum((nan_frac >= 0.5) & (nan_frac < 0.9)))
print("Columns with >90% NaN:", np.sum(nan_frac >= 0.9))

Columns with 0% NaN: 82
Columns with <5% NaN: 33
Columns with 5-50% NaN: 59
Columns with 50-90% NaN: 48
Columns with >90% NaN: 99


### 2.2 Drop non-predictive administrative columns

In [5]:
with open("dataset/x_train.csv", "r") as f:
    header = f.readline().strip()
feature_names = header.split(",")
feature_names = feature_names[1:]  # drop 'Id'
print(len(feature_names))
print(feature_names[:5])

321
['_STATE', 'FMONTH', 'IDATE', 'IMONTH', 'IDAY']


In [6]:
admin_cols = ['_STATE', 'FMONTH', 'IDATE', 'IMONTH', 'IDAY', 'IYEAR', 
              'DISPCODE', 'SEQNO', '_PSU', 'CTELENUM', 'PVTRESD1', 'COLGHOUS',
              'STATERES', 'CELLFON3', 'LADULT', 'NUMADULT', 'NUMMEN', 'NUMWOMEN',
              'CTELNUM1', 'CELLFON2', 'CADULT', 'PVTRESD2', 'CCLGHOUS', 'CSTATE',
              'LANDLINE', 'HHADULT', 'QSTVER', 'QSTLANG', 'MSCODE',
              '_STSTR', '_STRWT', '_RAWRAKE', '_WT2RAKE', '_CLLCPWT', '_DUALUSE',
              '_DUALCOR', '_LLCPWT']


admin_idx = [feature_names.index(c) for c in admin_cols if c in feature_names]
print("Dropping", len(admin_idx), "admin columns")

keep_mask = np.ones(len(feature_names), dtype=bool)
keep_mask[admin_idx] = False

x_train_clean = x_train[:, keep_mask]
x_test_clean = x_test[:, keep_mask]
feature_names_clean = [f for f, k in zip(feature_names, keep_mask) if k]

print(x_train_clean.shape)
print(x_test_clean.shape)
print(len(feature_names_clean))

Dropping 37 admin columns
(328135, 284)
(109379, 284)
284


### 2.3 Convert BRFSS "missing" sentinel codes to real NaN

This is the critical BRFSS-specific step. In this survey, missingness is usually hidden as specific numeric codes rather than true NaN:

Categorical variables: 7 = "Don't know/Not sure", 9 = "Refused"
Two-digit variables: 77 = "Don't know", 99 = "Refused"
Continuous variables (like PHYSHLTH, MENTHLTH): 77/88/99 or 777/999/7777/9999 depending on scale
77/99 at boundary of the variable's natural range is the giveaway — e.g. if a column's real values are 1–30 but it also has 77 and 99, those are sentinel codes, not real data points.

Rather than hardcoding all ~230 variable-specific rules (the official codebook has each one individually), here's a practical heuristic that works well for BRFSS: for each column, if the max value is one of {7, 9, 77, 88, 99, 777, 999, 7777, 9999} and it sits far above the bulk of the other values (i.e., it's a clear outlier code, not a natural continuation of the scale), treat it as missing.

In [7]:
sentinel_codes = {7, 9, 77, 88, 99, 777, 999, 7777, 9999}

def replace_sentinels(x, col_idx, verbose=False):
    col = x[:, col_idx].copy()
    non_nan = col[~np.isnan(col)]
    if len(non_nan) == 0:
        return col
    for code in sentinel_codes:
        if code in non_nan:
            # only treat as sentinel if it's clearly separated from the rest
            rest = non_nan[non_nan != code]
            if len(rest) > 0 and code > np.percentile(rest, 99) + 1:
                col[col == code] = np.nan
                if verbose:
                    print(f"Column {col_idx}: replaced code {code}")
    return col

x_train_sentinel = x_train_clean.copy()
for i in range(x_train_sentinel.shape[1]):
    x_train_sentinel[:, i] = replace_sentinels(x_train_sentinel, i)

# Check the impact on train:
before = np.isnan(x_train_clean).sum()
after = np.isnan(x_train_sentinel).sum()
print("NaNs before:", before)
print("NaNs after:", after)
print("Newly identified missing values:", after - before)


NaNs before: 43341504
NaNs after: 44757010
Newly identified missing values: 1415506


In [8]:
# Apply the identical replacement to test.
def get_sentinel_codes_for_col(x_train, col_idx):
    col = x_train[:, col_idx]
    non_nan = col[~np.isnan(col)]
    found = set()
    if len(non_nan) == 0:
        return found
    for code in sentinel_codes:
        if code in non_nan:
            rest = non_nan[non_nan != code]
            if len(rest) > 0 and code > np.percentile(rest, 99) + 1:
                found.add(code)
    return found

# Determine sentinel codes using x_train only
col_sentinels = [get_sentinel_codes_for_col(x_train_clean, i) for i in range(x_train_clean.shape[1])]

# Apply the same codes to test
x_test_sentinel = x_test_clean.copy()
for i, codes in enumerate(col_sentinels):
    for code in codes:
        x_test_sentinel[x_test_clean[:, i] == code, i] = np.nan

print("Test NaNs before:", np.isnan(x_test_clean).sum())
print("Test NaNs after:", np.isnan(x_test_sentinel).sum())

Test NaNs before: 14446569
Test NaNs after: 14919295


### 2.4 Drop columns with excessive missingness

In [9]:
#recompute missingness fractions on the updated arrays
nan_frac = np.isnan(x_train_sentinel).sum(axis=0) / x_train_sentinel.shape[0]

# Recheck distribution
print("0% NaN:", np.sum(nan_frac == 0))
print("<5% NaN:", np.sum((nan_frac > 0) & (nan_frac < 0.05)))
print("5-50% NaN:", np.sum((nan_frac >= 0.05) & (nan_frac < 0.5)))
print("50-90% NaN:", np.sum((nan_frac >= 0.5) & (nan_frac < 0.9)))
print(">90% NaN:", np.sum(nan_frac >= 0.9))

0% NaN: 17
<5% NaN: 60
5-50% NaN: 70
50-90% NaN: 41
>90% NaN: 96


In [10]:
# look at which columns have >90% missing values:
high_missing_idx = np.where(nan_frac >= 0.9)[0]
high_missing_names = [feature_names_clean[i] for i in high_missing_idx]
print(len(high_missing_names))
print(high_missing_names)

96
['NUMPHON2', 'INSULIN', 'BLDSUGAR', 'FEETCHK2', 'DOCTDIAB', 'CHKHEMO3', 'FEETCHK', 'EYEEXAM', 'DIABEYE', 'DIABEDU', 'CRGVREL1', 'CRGVLNG1', 'CRGVHRS1', 'CRGVPRB1', 'CRGVPERS', 'CRGVHOUS', 'CRGVMST2', 'VIDFCLT2', 'VIREDIF3', 'VIPRFVS2', 'VINOCRE2', 'VIEYEXM2', 'VIINSUR2', 'VICTRCT4', 'VIGLUMA2', 'VIMACDG2', 'CDHOUSE', 'CDASSIST', 'CDHELP', 'CDSOCIAL', 'CDDISCUS', 'WTCHSALT', 'LONGWTCH', 'DRADVISE', 'ASTHMAGE', 'ASATTACK', 'ASERVIST', 'ASDRVIST', 'ASRCHKUP', 'ASACTLIM', 'ASYMPTOM', 'ASNOSLEP', 'ASTHMED3', 'ASINHALR', 'HAREHAB1', 'STREHAB1', 'CVDASPRN', 'ASPUNSAF', 'RLIVPAIN', 'RDUCHART', 'RDUCSTRK', 'ARTTODAY', 'ARTHWGT', 'ARTHEXER', 'ARTHEDU', 'TETANUS', 'HPVADVC2', 'HPVADSHT', 'SHINGLE2', 'HADMAM', 'HOWLONG', 'HADPAP2', 'LASTPAP2', 'HPVTEST', 'HPLSTTST', 'HADHYST2', 'PROFEXAM', 'LENGEXAM', 'LSTBLDS3', 'HADSGCO1', 'LASTSIG3', 'PCPSAAD2', 'PCPSADI1', 'PCPSARE1', 'PSATEST1', 'PSATIME', 'PCPSARS1', 'PCPSADE1', 'PCDMDECN', 'SCNTPAID', 'SCNTWRK1', 'SCNTLPAD', 'SCNTLWK1', 'CASTHNO2', 'EMTS

In [11]:
# Drop all 96 high-missing columns (None of these are core cardiovascular risk factors, they're mostly downstream management/detail questions for people who already have a specific condition.): 
high_missing_idx_set = set(high_missing_idx)  # from before, based on nan_frac >= 0.9

keep_mask2 = np.ones(x_train_sentinel.shape[1], dtype=bool)
keep_mask2[list(high_missing_idx_set)] = False

x_train_v2 = x_train_sentinel[:, keep_mask2]
x_test_v2 = x_test_sentinel[:, keep_mask2]
feature_names_v2 = [f for f, k in zip(feature_names_clean, keep_mask2) if k]

print(x_train_v2.shape)
print(x_test_v2.shape)
print(len(feature_names_v2))

(328135, 188)
(109379, 188)
188


### 2.5 Impute remaining missing values:

In [12]:
nan_frac_v2 = np.isnan(x_train_v2).sum(axis=0) / x_train_v2.shape[0]
print("Still have NaNs in", np.sum(nan_frac_v2 > 0), "out of", x_train_v2.shape[1], "columns")
print("Max remaining missing fraction:", nan_frac_v2.max())

Still have NaNs in 171 out of 188 columns
Max remaining missing fraction: 0.8853002575159614


In [13]:
#Look at remaining data with >50% missing values:
mid_missing_idx = np.where((nan_frac_v2 >= 0.5) & (nan_frac_v2 < 0.9))[0]
mid_missing_names = [feature_names_v2[i] for i in mid_missing_idx]
for name, frac in sorted(zip(mid_missing_names, nan_frac_v2[mid_missing_idx]), key=lambda x: -x[1]):
    print(f"{name:15s} {frac:.4f}")

CASTHDX2        0.8853
BLDSTOOL        0.8785
HADSIGM3        0.8781
DIABAGE2        0.8713
RCSGENDR        0.8707
ASTHNOW         0.8701
RCSRLTN2        0.8697
STOPSMK2        0.8614
_CHISPNC        0.8593
_CRACE1         0.8590
_CPRACE         0.8590
PREGNANT        0.8534
SCNTMNY1        0.8427
SCNTMEL1        0.8345
CRGVEXPT        0.8226
PDIABTST        0.8214
PREDIAB1        0.8134
CAREGIV1        0.7544
WHRTST10        0.7473
HIVTSTD3        0.7418
CIMEMLOS        0.7383
LASTSMK2        0.7253
JOINPAIN        0.7040
_PNEUMO2        0.6966
ARTHDIS2        0.6955
LMTJOIN3        0.6952
ARTHSOCL        0.6951
_FLSHOT6        0.6868
SXORIENT        0.6281
TRNSGNDR        0.6272
BPMEDS          0.5990
SMOKDAY2        0.5840
FLSHTMY2        0.5692
IMFVPLAC        0.5684
PADUR2_         0.5627
PAFREQ2_        0.5575
EXEROFT2        0.5574
EXERHMM2        0.5538
MAXDRNKS        0.5375
AVEDRNK2        0.5288
DRNK3GE5        0.5232


In [14]:
# Drop all features with > 50% missing values. Might be worth adding BPMEDS, AVEDRNK2, MAXDRNKS, DRNK3GE5 later to improve model if necessary
mid_missing_idx_set = set(mid_missing_idx)  # from before

keep_mask3 = np.ones(x_train_v2.shape[1], dtype=bool)
keep_mask3[list(mid_missing_idx_set)] = False

x_train_v3 = x_train_v2[:, keep_mask3]
x_test_v3 = x_test_v2[:, keep_mask3]
feature_names_v3 = [f for f, k in zip(feature_names_v2, keep_mask3) if k]

print(x_train_v3.shape)
print(x_test_v3.shape)
print(len(feature_names_v3))

(328135, 147)
(109379, 147)
147


In [15]:
# impute median if Nan value:
## worth trying other imputation techniques later on to improve outcome
nan_frac_v3 = np.isnan(x_train_v3).sum(axis=0) / x_train_v3.shape[0]
print("Columns with NaNs:", np.sum(nan_frac_v3 > 0), "out of", x_train_v3.shape[1])
print("Max remaining missing fraction:", nan_frac_v3.max())

col_medians = np.nanmedian(x_train_v3, axis=0)
print("Columns with NaN median (fully empty):", np.sum(np.isnan(col_medians)))

def impute_with_medians(x, medians):
    x_out = x.copy()
    inds = np.where(np.isnan(x_out))
    x_out[inds] = np.take(medians, inds[1])
    return x_out

x_train_imputed = impute_with_medians(x_train_v3, col_medians)
x_test_imputed = impute_with_medians(x_test_v3, col_medians)

print("Remaining NaNs in train:", np.isnan(x_train_imputed).sum())
print("Remaining NaNs in test:", np.isnan(x_test_imputed).sum())

Columns with NaNs: 130 out of 147
Max remaining missing fraction: 0.4898806893504198
Columns with NaN median (fully empty): 0
Remaining NaNs in train: 0
Remaining NaNs in test: 0


### 2.6 Remove near-constant (zero/low-variance) columns:

A column where almost every respondent has the same value carries little to no predictive power, but still adds noise and computation.

In [16]:
col_std = np.std(x_train_imputed, axis=0)
print("Columns with std == 0 (constant):", np.sum(col_std == 0))
print("Columns with very low std (< 0.01):", np.sum(col_std < 0.01))

# Look at the lowest-variance columns
low_var_idx = np.argsort(col_std)[:15]
for i in low_var_idx:
    print(f"{feature_names_v3[i]:15s} std={col_std[i]:.6f}")

Columns with std == 0 (constant): 0
Columns with very low std (< 0.01): 0
_VEG23          std=0.010901
_FRT16          std=0.014604
HTM4            std=0.103858
NUMHHOL2        std=0.170973
CHCKIDNY        std=0.183797
CVDSTRK3        std=0.197672
DIFFDRES        std=0.201203
BLIND           std=0.212923
PAMISS1_        std=0.214419
_RFSEAT2        std=0.215781
_RFDRHV5        std=0.218123
_HCVU651        std=0.250096
HLTHPLN1        std=0.259164
DIFFALON        std=0.263813
CHCCOPD1        std=0.270987


None of these should be dropped: low variance due to genuine class imbalance (rare condition) is very different from low variance due to a column being uninformative/constant.

### 2.7 Check correlation btw. features

In [17]:
# Compute correlation matrix (147x147)
corr_matrix = np.corrcoef(x_train_imputed, rowvar=False)

# Find pairs with |correlation| > 0.9, excluding the diagonal
high_corr_pairs = []
n = corr_matrix.shape[0]
for i in range(n):
    for j in range(i+1, n):
        if abs(corr_matrix[i, j]) > 0.9:
            high_corr_pairs.append((feature_names_v3[i], feature_names_v3[j], corr_matrix[i, j]))

print(f"Found {len(high_corr_pairs)} highly correlated pairs")
for a, b, c in sorted(high_corr_pairs, key=lambda x: -abs(x[2])):
    print(f"{a:15s} <-> {b:15s}  corr={c:.3f}")

Found 40 highly correlated pairs
TOLDHI2         <-> _RFCHOL          corr=-1.000
ASTHMA3         <-> _LTASTH1         corr=-1.000
HAVARTH3        <-> _DRDXAR1         corr=1.000
EXERANY2        <-> _TOTINDA         corr=1.000
HIVTST6         <-> _AIDTST3         corr=1.000
MAXVO2_         <-> FC60_            corr=1.000
_VEGRESP        <-> _VEGETEX         corr=-0.998
_PRACE1         <-> _MRACE1          corr=0.995
_FRTRESP        <-> _FRUITEX         corr=-0.995
HTIN4           <-> HTM4             corr=0.994
EXERHMM1        <-> PADUR1_          corr=0.993
BPHIGH4         <-> _RFHYPE5         corr=-0.991
EDUCA           <-> _EDUCAG          corr=0.988
_AGEG5YR        <-> _AGE80           corr=0.977
_AGE80          <-> _AGE_G           corr=0.971
_AGE80          <-> FC60_            corr=-0.969
_AGE80          <-> MAXVO2_          corr=-0.969
_MISFRTN        <-> _FRTRESP         corr=-0.964
BLOODCHO        <-> _CHOLCHK         corr=0.963
_MISFRTN        <-> _FRUITEX         corr=0.960

In [18]:
# Build a graph of highly correlated features and find connected groups

graph = defaultdict(set)
for a, b, c in high_corr_pairs:
    graph[a].add(b)
    graph[b].add(a)

visited = set()
groups = []

def dfs(node, group):
    visited.add(node)
    group.append(node)
    for neighbor in graph[node]:
        if neighbor not in visited:
            dfs(neighbor, group)

for node in graph:
    if node not in visited:
        group = []
        dfs(node, group)
        groups.append(group)

print(f"Found {len(groups)} correlated groups")
for g in groups:
    print(g)

Found 18 correlated groups
['HLTHPLN1', '_HCVU651']
['BPHIGH4', '_RFHYPE5']
['BLOODCHO', '_CHOLCHK']
['TOLDHI2', '_RFCHOL']
['ASTHMA3', '_ASTHMS1', '_LTASTH1', '_CASTHM1']
['HAVARTH3', '_DRDXAR1']
['EDUCA', '_EDUCAG']
['EXERANY2', '_TOTINDA']
['EXERHMM1', 'PADUR1_']
['HIVTST6', '_AIDTST3']
['_PRACE1', '_MRACE1']
['_AGEG5YR', '_AGE_G', 'FC60_', 'MAXVO2_', '_AGE80']
['HTIN4', 'HTM4']
['DROCDY3_', '_DRNKWEK']
['_MISFRTN', '_MISVEGN', '_VEGRESP', '_VEGETEX', '_FRTRESP', '_FRUITEX']
['_MINAC11', 'PAMIN11_']
['_MINAC21', 'PAMIN21_']
['_PACAT1', '_PAINDX1', '_PAREC1', '_PA150R2', '_PA300R2']


In [19]:
name_to_idx = {name: i for i, name in enumerate(feature_names_v3)}
age_fitness_group = ['_AGEG5YR', '_AGE_G', 'MAXVO2_', 'FC60_', '_AGE80']
for name in age_fitness_group:
    print(f"{name:12s} missing_frac={nan_frac_v3[name_to_idx[name]]:.4f}")

_AGEG5YR     missing_frac=0.0000
_AGE_G       missing_frac=0.0000
MAXVO2_      missing_frac=0.0120
FC60_        missing_frac=0.0120
_AGE80       missing_frac=0.0000


In [20]:
# Define which features to drop based on correlation: keep only features with less missing values at the beginning in each correlation group
name_to_idx = {name: i for i, name in enumerate(feature_names_v3)}
forced_keep = {'_AGEG5YR'}

to_drop = set()
for g in groups:
    if any(name in forced_keep for name in g):
        best = next(name for name in g if name in forced_keep)
    else:
        best = min(g, key=lambda name: nan_frac_v3[name_to_idx[name]])
    for name in g:
        if name != best:
            to_drop.add(name)

print(f"Dropping {len(to_drop)} redundant columns")
print(sorted(to_drop))

Dropping 30 redundant columns
['FC60_', 'HTIN4', 'MAXVO2_', 'PADUR1_', 'PAMIN11_', 'PAMIN21_', '_AGE80', '_AGE_G', '_AIDTST3', '_ASTHMS1', '_CASTHM1', '_CHOLCHK', '_DRDXAR1', '_DRNKWEK', '_EDUCAG', '_FRTRESP', '_FRUITEX', '_HCVU651', '_LTASTH1', '_MISVEGN', '_MRACE1', '_PA150R2', '_PA300R2', '_PACAT1', '_PAREC1', '_RFCHOL', '_RFHYPE5', '_TOTINDA', '_VEGETEX', '_VEGRESP']


In [21]:
# Drop features
drop_idx = [name_to_idx[name] for name in to_drop]

keep_mask4 = np.ones(x_train_imputed.shape[1], dtype=bool)
keep_mask4[drop_idx] = False

x_train_v4 = x_train_imputed[:, keep_mask4]
x_test_v4 = x_test_imputed[:, keep_mask4]
feature_names_v4 = [f for f, k in zip(feature_names_v3, keep_mask4) if k]

print(x_train_v4.shape)
print(x_test_v4.shape)
print(len(feature_names_v4))

(328135, 117)
(109379, 117)
117


### 2.8 Standardization

In [22]:
mu = np.mean(x_train_v4, axis=0)
sigma = np.std(x_train_v4, axis=0)

# safety check: avoid divide-by-zero for any accidentally-constant column
sigma[sigma == 0] = 1

x_train_final = (x_train_v4 - mu) / sigma
x_test_final = (x_test_v4 - mu) / sigma  # use TRAIN's mu/sigma, not test's own!

print(x_train_final.shape)
print(x_test_final.shape)
print("Mean check (should be ~0):", np.mean(x_train_final, axis=0)[:5])
print("Std check (should be ~1):", np.std(x_train_final, axis=0)[:5])

(328135, 117)
(109379, 117)
Mean check (should be ~0): [-1.46207645e-16 -7.24975111e-17  2.07531704e-16  2.54477524e-16
  1.24207205e-16]
Std check (should be ~1): [1. 1. 1. 1. 1.]


### 2.9 Check class imbalance

In [23]:
print("Fraction y=1 (heart attack):", np.mean(y_train == 1))
print("Fraction y=-1 (no heart attack):", np.mean(y_train == -1))

Fraction y=1 (heart attack): 0.08830207079403295
Fraction y=-1 (no heart attack): 0.911697929205967


Highly imbalanced data set: A trivial model that always predicts "no heart attack" would already get 91.2% accuracy. Accuracy alone is a misleading metric here. Look at metrics like F1-score, precision, recall, or balanced accuracy to know if model is actually learning to detect the minority class (heart attack cases), not just exploiting the imbalance.

# Save cleaned arrays to disk to not have to re-run whole pipeline every time. Remove before submitting !!!!

In [24]:
np.save("x_train_final.npy", x_train_final)
np.save("x_test_final.npy", x_test_final)
np.save("y_train.npy", y_train)
np.save("feature_names_v4.npy", np.array(feature_names_v4))


In [25]:
test_ids = np.genfromtxt("dataset/x_test.csv", delimiter=",", skip_header=1, usecols=0, dtype=int)
np.save("test_ids.npy", test_ids)

# 3. Model implementation

In [26]:
import numpy as np
from helpers import *
from collections import defaultdict
from implementations import *
import os


In [27]:

x_train_final = np.load("x_train_final.npy")
x_test_final = np.load("x_test_final.npy")
y_train = np.load("y_train.npy")
feature_names_v4 = np.load("feature_names_v4.npy", allow_pickle=True)  # array of strings needs this flag sometimes
test_ids = np.load("test_ids.npy")

In [28]:
print(np.unique(y_train))
# convert current output of y_train as [0, 1] => needed fo log. regression and consistent for all models

y01 = (y_train + 1) / 2
print(np.unique(y01))                      # [0. 1.]
print("positive rate:", y01.mean())        # fraction of heart attack cases

tx = np.c_[np.ones(len(x_train_final)), x_train_final]          # (N, D)
tx_test = np.c_[np.ones(len(x_test_final)), x_test_final]       # (N_test, D)
N, D = tx.shape
print(tx.shape, tx_test.shape)


[-1  1]
[0. 1.]
positive rate: 0.08830207079403295
(328135, 118) (109379, 118)


## 3.1 Linear regression using gradient descent

In [29]:
#Step 1: build tx and chose parameters => compute a safe gamma: GD on MSE is stable only if gamma < 2/L, where L is the largest eigenvalue of X^T X / N

L = np.linalg.eigvalsh(tx.T @ tx / N).max()   # largest eigenvalue
gamma = 1.0 / L
print("L =", L, " gamma =", gamma)

# find out which max_iters to use by checking difference with least_squares
# Reference: closed-form solution
w_ls, loss_ls = least_squares(y01, tx)
print("least squares loss:", loss_ls)

# GD with an increasing number of iterations
for iters in [50, 200, 1000, 3000]:
    w_gd, loss_gd = mean_squared_error_gd(y01, tx, np.zeros(D), iters, gamma)
    print(iters, "iters -> loss:", loss_gd, " gap to LS:", loss_gd - loss_ls)

# chose 200 iterations as going from 200 to 3000 iterations improves the loss by about 5e-7, while increasing run time a lot

L = 8.873852418142327  gamma = 0.11269062779943576
least squares loss: 0.03359143892710773
50 iters -> loss: 0.033597802469880374  gap to LS: 6.363542772641617e-06
200 iters -> loss: 0.033592116850207264  gap to LS: 6.779230995318519e-07
1000 iters -> loss: 0.03359169858127964  gap to LS: 2.596541719065981e-07
3000 iters -> loss: 0.033591599950941076  gap to LS: 1.610238333435432e-07


<div style = "white-space: pre-line">Performance:
(328135, 118)
L = 8.923755473320941  gamma = 0.11206044394533973
least squares loss: 0.03359152416484328
50 iters -> loss: 0.033597955972728785  gap to LS: 6.431807885502738e-06
200 iters -> loss: 0.03359220922057452  gap to LS: 6.850557312393368e-07
1000 iters -> loss: 0.033591784761033784  gap to LS: 2.605961905016718e-07
3000 iters -> loss: 0.0335916857695453  gap to LS: 1.616047020161493e-07
</div>

In [30]:
# Functions needed for Step 2
def train_linreg_gd(x, y, fold):
    """Train linear regression with GD on the training rows, return the weights w."""
    w, _ = mean_squared_error_gd(y, x, np.zeros(x.shape[1]), max_iterations_lr, gamma)
    return w

def score_linear(w, x):
    """Raw score of each row (not a probability)."""
    return x @ w

In [31]:
# Step 2: 5-fold cross validation (shared cross_validate from implementations.py)
thresholds = np.linspace(0.05, 0.5, 46)
max_iterations_lr = 200         # from Step 1

t_gd, m_gd, s_gd = cross_validate(train_linreg_gd, score_linear, tx, y01, thresholds)


fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.21 | F1 0.4106 ± 0.0041 | P 0.3322 ± 0.0032 | R 0.5374 ± 0.0060


<div style = "white-space: pre-line">Performance:
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.21 | F1 0.4106 ± 0.0042 | P 0.3322 ± 0.0033 | R 0.5375 ± 0.0061 </div>

In [32]:
# # Final model for Linear regression using gradient descent: train on ALL training rows (tx already has the bias column)
max_iters_linreg_gd = 200
w_final, loss_final = mean_squared_error_gd(y01, tx, np.zeros(tx.shape[1]), max_iters_linreg_gd, gamma)
print("final train loss:", loss_final)

final train loss: 0.033592116850207264


<div style = "white-space: pre-line">Performance:
final train loss: 0.03359220922057452 </div>

In [33]:
# prepare the test set and predict

best_t = t_gd                                   # 0.21, chosen by CV

scores_test = tx_test @ w_final                                # raw scores
y_pred_test = np.where(scores_test > best_t, 1, -1)            # back to {-1, 1}

# Sanity check: the share of predicted positives should be similar on train and test
print("predicted positive rate, test: ", np.mean(y_pred_test == 1))
print("predicted positive rate, train:", np.mean((tx @ w_final) > best_t))

predicted positive rate, test:  0.14302562649137404
predicted positive rate, train: 0.14288021698386336


<div style = "white-space: pre-line">Performance:
predicted positive rate, test:  0.14300734144579855
predicted positive rate, train: 0.14288935956237525 </div>

In [34]:
# submission file for Linear regression using gradient descent: 
os.makedirs("submissions", exist_ok=True)
create_csv_submission(test_ids, y_pred_test, "submissions/submission_linreg_gd.csv")

## 3.2 Linear regression using stochastic gradient descent

In [35]:
# Step 1: choose gamma and max_iters:

# One stratified split: fold 0 = validation, the other 4 folds = training
folds = stratified_kfold_indices(y01, 5, 42)
val_idx = folds[0]
tr_idx = np.concatenate(folds[1:])

thresholds = np.linspace(0.05, 0.5, 46)


max_iters_sgd = 300000      # roughly one pass over the training rows
for g in [1e-6, 3e-6, 1e-5, 3e-5]:
    for iters in [300000, 600000]:
        np.random.seed(0)
        w, loss = mean_squared_error_sgd(y01[tr_idx], tx[tr_idx],
                                         np.zeros(tx.shape[1]), iters, g)
        f1 = best_f1(tx[val_idx] @ w, y01[val_idx], thresholds)
        print("gamma", g, "iters", iters, "-> loss", round(loss, 5), " val F1", round(f1, 4))

# gamma = 1e-5 is the safest choice: it is in the middle of the plateau, has the lowest loss, and is well below the divergence region (1e-3)

gamma 1e-06 iters 300000 -> loss 0.03676  val F1 0.3889
gamma 1e-06 iters 600000 -> loss 0.03526  val F1 0.4023
gamma 3e-06 iters 300000 -> loss 0.03451  val F1 0.4096
gamma 3e-06 iters 600000 -> loss 0.03379  val F1 0.4149
gamma 1e-05 iters 300000 -> loss 0.03366  val F1 0.4144
gamma 1e-05 iters 600000 -> loss 0.03364  val F1 0.4154
gamma 3e-05 iters 300000 -> loss 0.03367  val F1 0.4145
gamma 3e-05 iters 600000 -> loss 0.03368  val F1 0.4159


<div style = "white-space: pre-line">Performance:
gamma 1e-06 iters 300000 -> loss 0.03676  val F1 0.3887
gamma 1e-06 iters 600000 -> loss 0.03526  val F1 0.4021
gamma 3e-06 iters 300000 -> loss 0.03451  val F1 0.4095
gamma 3e-06 iters 600000 -> loss 0.03379  val F1 0.4145
gamma 1e-05 iters 300000 -> loss 0.03366  val F1 0.4144
gamma 1e-05 iters 600000 -> loss 0.03364  val F1 0.415
gamma 3e-05 iters 300000 -> loss 0.03367  val F1 0.4143
gamma 3e-05 iters 600000 -> loss 0.03368  val F1 0.4161 </div>

In [36]:
def train_linreg_sgd(x, y, fold):
    """Train linear regression with SGD (batch size 1) on the training rows, return the weights w.
    The fold number is used as seed, so that the SGD sampling is reproducible."""
    np.random.seed(fold)
    w, _ = mean_squared_error_sgd(y, x, np.zeros(x.shape[1]), max_iters_sgd, gamma_sgd)
    return w

def score_linear(w, x):
    """SAME AS FOR 3.1 LR WITH GRADIENT DESCENT
    Raw score of each row (not a probability)."""
    return x @ w



In [37]:
# Step 2: 5 fold cross validation with chosen parameters to select best w
gamma_sgd = 1e-5                # from Step 1
max_iters_sgd = 600000          # from Step 1

t_sgd, m_sgd, s_sgd = cross_validate(train_linreg_sgd, score_linear, tx, y01, thresholds)

fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.20 | F1 0.4099 ± 0.0038 | P 0.3214 ± 0.0053 | R 0.5659 ± 0.0115


<div style = "white-space: pre-line">Performance:
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.21 | F1 0.4099 ± 0.0037 | P 0.3329 ± 0.0059 | R 0.5336 ± 0.0125 </div>

In [38]:
# Step 3: final model on training data and test predictions
np.random.seed(0)
w_sgd, loss_sgd = mean_squared_error_sgd(y01, tx, np.zeros(tx.shape[1]), max_iters_sgd, gamma_sgd)
best_t_sgd = t_sgd

scores_test = tx_test @ w_sgd
y_pred_sgd = np.where(scores_test > best_t_sgd, 1, -1)

print("final train loss:", loss_sgd)                      # should be ~0.0336
print("predicted positive rate, test: ", np.mean(y_pred_sgd == 1))
print("predicted positive rate, train:", np.mean((tx @ w_sgd) > best_t_sgd))


final train loss: 0.03361855521135971
predicted positive rate, test:  0.15763537790617943
predicted positive rate, train: 0.15737425145138434


<div style = "white-space: pre-line">Performance:
final train loss: 0.0336186891994748
predicted positive rate, test:  0.14392159372457236
predicted positive rate, train: 0.14372438173312813 </div>

In [39]:
#create file for submission 
os.makedirs("submissions", exist_ok=True)
create_csv_submission(test_ids, y_pred_sgd, "submissions/submission_linreg_sgd.csv")

## 3.3 Decision tree


In [40]:
# Step 1: bin the features to make run faster
# test set must be binned with edges from training data!!  
# functions in implementations.py

In [41]:
# Step 2: Build the tree
## functions in implements.py

In [42]:
#Step3: select hyperparameters (based on one fold to make it faster)
edges = make_edges(x_train_final)
B = bin_data(x_train_final, edges)
n_bins = max(len(e) for e in edges) + 1

folds = stratified_kfold_indices(y01, 5, 42)
val_idx, tr_idx = folds[0], np.concatenate(folds[1:])

tree = build_tree(B, y01, tr_idx, n_bins, 0, max_depth=6, min_leaf=200)
scores = predict_tree(tree, B[val_idx])
print("val F1:", best_f1(scores, y01[val_idx], np.linspace(0.05, 0.7, 66)))

val F1: 0.3887278365805973


<div style = "white-space: pre-line">Performance:
val F1: 0.3887278365805973 </div>

In [43]:
thr = np.linspace(0.05, 0.7, 66)
for depth in [3, 5, 8, 10]:
    for leaf in [50, 200, 1000]:
        tree = build_tree(B, y01, tr_idx, n_bins, 0, max_depth=depth, min_leaf=leaf)
        f1 = best_f1(predict_tree(tree, B[val_idx]), y01[val_idx], thr)
        print("depth", depth, "min_leaf", leaf, "-> val F1", round(f1, 4))
# depth 8 is best (F1 climbs and then drops slightly) 50 ~200 for min_leaf

depth 3 min_leaf 50 -> val F1 0.34
depth 3 min_leaf 200 -> val F1 0.34
depth 3 min_leaf 1000 -> val F1 0.34
depth 5 min_leaf 50 -> val F1 0.3837
depth 5 min_leaf 200 -> val F1 0.3838
depth 5 min_leaf 1000 -> val F1 0.3796
depth 8 min_leaf 50 -> val F1 0.4052
depth 8 min_leaf 200 -> val F1 0.4054
depth 8 min_leaf 1000 -> val F1 0.3984
depth 10 min_leaf 50 -> val F1 0.4021
depth 10 min_leaf 200 -> val F1 0.4037
depth 10 min_leaf 1000 -> val F1 0.3987


<div style = "white-space: pre-line">Performance:
depth 3 min_leaf 50 -> val F1 0.34
depth 3 min_leaf 200 -> val F1 0.34
depth 3 min_leaf 1000 -> val F1 0.34
depth 5 min_leaf 50 -> val F1 0.3837
depth 5 min_leaf 200 -> val F1 0.3838
depth 5 min_leaf 1000 -> val F1 0.3796
depth 8 min_leaf 50 -> val F1 0.4052
depth 8 min_leaf 200 -> val F1 0.4054
depth 8 min_leaf 1000 -> val F1 0.3984
depth 10 min_leaf 50 -> val F1 0.4021
depth 10 min_leaf 200 -> val F1 0.4037
depth 10 min_leaf 1000 -> val F1 0.3987 </div>


In [44]:
# Functions needed for Step 4
def train_tree(Bx, y, _fold):
    """Build a tree on the binned training rows.
    _fold is not used (tree building is deterministic), it is only there to match cross_validate."""
    return build_tree(Bx, y, np.arange(len(y)), n_bins, 0, max_depth_tree, min_leaf_tree)

def score_tree(tree, Bx):
    """Fraction of positives in the leaf of each row."""
    return predict_tree(tree, Bx)

In [45]:
#Step 4: Run 5-fold CV with chosen settings: 
# Step 4: 5-fold CV with the chosen settings

max_depth_tree = 8           #from last steps
min_leaf_tree = 200      #from last steps
thr = np.linspace(0.05, 0.7, 66)

t_tree, m_tree, s_tree = cross_validate(train_tree, score_tree, B, y01, thr)

fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.20 | F1 0.4050 ± 0.0016 | P 0.3232 ± 0.0012 | R 0.5421 ± 0.0053


<div style = "white-space: pre-line"> Performance:
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.20 | F1 0.4050 ± 0.0016 | P 0.3232 ± 0.0012 | R 0.5421 ± 0.0053 </div>

So far: 
Linear regression, GD -> CV F1 = 0.4106
Linear regression, SGD -> CV F1 = 0.4099
Decision tree (depth 8, min_leaf 200) -> CV = 0.4050

=> The tree is about 0.006 below linear regression -> a single tree doesn't beat linear model on this data.

In [46]:
# Final tree: train on ALL training rows
all_idx = np.arange(len(y01))
tree_final = build_tree(B, y01, all_idx, n_bins, 0, max_depth=8, min_leaf=200)

# Bin the test set with the TRAIN edges (not its own)
B_test = bin_data(x_test_final, edges)
scores_test = predict_tree(tree_final, B_test)

best_t_tree  = t_tree
y_pred_tree = np.where(scores_test > best_t_tree, 1, -1)

print("predicted positive rate, test: ", np.mean(y_pred_tree == 1))
print("predicted positive rate, train:", np.mean(predict_tree(tree_final, B) > best_t_tree))

os.makedirs("submissions", exist_ok=True)
create_csv_submission(test_ids, y_pred_tree, "submissions/submission_tree.csv")

predicted positive rate, test:  0.1547737682736174
predicted positive rate, train: 0.15493013546253828


<div style = "white-space: pre-line">Performance:
predicted positive rate, test:  0.1547737682736174
predicted positive rate, train: 0.15493013546253828 </div>

# 3.4 Least Squares #

In [47]:
# train LS on all training data
w_ls, loss_ls = least_squares(y01, tx)

# sanity check to compare loss with linear regression using gradient descent
# should be very similar losses as they both minimize the MSE, but in different ways
print("Loss:", loss_ls)
print ("w shape:", w_ls.shape)

Loss: 0.03359143892710773
w shape: (118,)


In [48]:
# get raw scores on training data
scores_ls = tx @ w_ls

# does NOT output probabilities, but rather raw scores of linear predictions
print(scores_ls[:10])
print(scores_ls.shape)
print(scores_ls.min())
print(scores_ls.max())

[-0.00925862  0.12629139  0.04777652 -0.00597018  0.09993692  0.18267053
  0.35112621  0.12364857  0.01593858  0.07653007]
(328135,)
-0.15002410363270985
0.7747685329171816


In [49]:
# LS training function to be used in cross validation
def train_ls(x_tr, y_tr, fold):
    """
    Train least-squares regression on the training fold.

    Parameters
    x_tr : training feature matrix
    y_tr : training labels in {0, 1}
    fold : fold number (required by cross_validate)

    Returns
    w : fitted least-squares weights
    """

    w, _ = least_squares(y_tr, x_tr)

    return w


In [50]:
def score_ls(w, x_val):
    """
    Generate raw least-squares scores on validation data.

    """

    scores = x_val @ w

    return scores

In [51]:
thresholds = np.linspace(0.05, 0.5, 46)

print("Thresholds:", thresholds)

Thresholds: [0.05 0.06 0.07 0.08 0.09 0.1  0.11 0.12 0.13 0.14 0.15 0.16 0.17 0.18
 0.19 0.2  0.21 0.22 0.23 0.24 0.25 0.26 0.27 0.28 0.29 0.3  0.31 0.32
 0.33 0.34 0.35 0.36 0.37 0.38 0.39 0.4  0.41 0.42 0.43 0.44 0.45 0.46
 0.47 0.48 0.49 0.5 ]


In [52]:
# 5-fold stratified cross validation
best_t_ls, mean_ls, std_ls = cross_validate(
    train_fn=train_ls,
    score_fn=score_ls,
    x=tx,
    y01=y01,
    thresholds=thresholds,
    k=5,
    seed=42
)

fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.21 | F1 0.4105 ± 0.0039 | P 0.3321 ± 0.0031 | R 0.5373 ± 0.0058


In [53]:
# check CV results
print("\nLeast Squares CV results:")
print("Best threshold:", best_t_ls)
print("Mean precision:", mean_ls[0])
print("Mean recall:", mean_ls[1])
print("Mean F1:", mean_ls[2])

print("\nStandard deviations:")
print("Precision std:", std_ls[0])
print("Recall std:", std_ls[1])
print("F1 std:", std_ls[2])


Least Squares CV results:
Best threshold: 0.21000000000000002
Mean precision: 0.33213816189010653
Mean recall: 0.5373252804141502
Mean F1: 0.4105178962054469

Standard deviations:
Precision std: 0.0030787316115222026
Recall std: 0.005755834268544541
F1 std: 0.00391252070567854


In [54]:
# now train least squares using ALL of the training data
w_ls_final, loss_ls_final = least_squares(y01, tx)

print("Final LS loss:", loss_ls_final)
print("w shape:", w_ls_final.shape)

Final LS loss: 0.03359143892710773
w shape: (118,)


In [55]:
# calculate raw LS scores with predefined test matrix (with bias column)
scores_test_ls = tx_test @ w_ls_final

print("Test scores shape:", scores_test_ls.shape)
print("Minimum score:", scores_test_ls.min())
print("Maximum score:", scores_test_ls.max())
print("First 10 scores:", scores_test_ls[:10])

Test scores shape: (109379,)
Minimum score: -0.13125370998167576
Maximum score: 0.7416411375664479
First 10 scores: [ 0.13513425 -0.07314583  0.07385477  0.06152322  0.05974348 -0.04200942
  0.02945641 -0.02290573  0.20850008  0.19340943]


In [56]:
# convert scores into predictions
y_pred_test_ls01 = (scores_test_ls > best_t_ls).astype(int)

print("First 20 predictions:", y_pred_test_ls01[:20])
print("Test positive rate:", np.mean(y_pred_test_ls01 == 1))
print("Test negative rate:", np.mean(y_pred_test_ls01 == 0))
print("Number of positive predictions:", np.sum(y_pred_test_ls01 == 1))
print("Number of negative predictions:", np.sum(y_pred_test_ls01 == 0))

# number of positive rate similar to linear regression with gradient descent! good sanity check

First 20 predictions: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]
Test positive rate: 0.1430164839685863
Test negative rate: 0.8569835160314138
Number of positive predictions: 15643
Number of negative predictions: 93736


In [57]:
# now convert back to labels -1, 1 - do they  need it in this format for the submission?
y_pred_test_ls = 2 * y_pred_test_ls01 - 1

print("First 20 submission predictions:", y_pred_test_ls[:20])
print("Unique labels:", np.unique(y_pred_test_ls))

First 20 submission predictions: [-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1  1]
Unique labels: [-1  1]


# do I need this? for the submission file - if so, change cell underneath to code cell

sub = np.genfromtxt(
    "dataset/sample_submission.csv",
    delimiter=",",
    skip_header=1
)

test_ids = sub[:, 0].astype(int)

print("\nNumber of test IDs:", len(test_ids))
print("Number of predictions:", len(y_pred_test_ls))


# Create submissions directory if it doesn't exist
import os

os.makedirs("submissions", exist_ok=True)


# Save submission
np.savetxt(
    "submissions/submission_linreg_ls.csv",
    np.c_[test_ids, y_pred_test_ls],
    delimiter=",",
    fmt="%d",
    header="Id,Prediction",
    comments=""
)

print("LS submission saved!")


# ------------------------------------------------------------
# 13. Verify submission
# ------------------------------------------------------------

print(
    np.genfromtxt(
        "submissions/submission_linreg_ls.csv",
        delimiter=",",
        skip_header=1
    )[:10]
)

# 3.5 Ridge Regression #

In [59]:
# sanity check first
lambdas = np.logspace(-6, 0, 13)

print(lambdas)

for lambda_ in lambdas:
    w_ridge, loss_ridge = ridge_regression(y01, tx, lambda_)

    print(
        f"lambda={lambda_}: "
        f"loss={loss_ridge:.6f}, "
        f"w shape={w_ridge.shape}"
    )

[1.00000000e-06 3.16227766e-06 1.00000000e-05 3.16227766e-05
 1.00000000e-04 3.16227766e-04 1.00000000e-03 3.16227766e-03
 1.00000000e-02 3.16227766e-02 1.00000000e-01 3.16227766e-01
 1.00000000e+00]
lambda=1e-06: loss=0.033591, w shape=(118,)
lambda=3.162277660168379e-06: loss=0.033591, w shape=(118,)
lambda=1e-05: loss=0.033591, w shape=(118,)
lambda=3.1622776601683795e-05: loss=0.033591, w shape=(118,)
lambda=0.0001: loss=0.033591, w shape=(118,)
lambda=0.00031622776601683794: loss=0.033591, w shape=(118,)
lambda=0.001: loss=0.033592, w shape=(118,)
lambda=0.0031622776601683794: loss=0.033592, w shape=(118,)
lambda=0.01: loss=0.033594, w shape=(118,)
lambda=0.03162277660168379: loss=0.033614, w shape=(118,)
lambda=0.1: loss=0.033755, w shape=(118,)
lambda=0.31622776601683794: loss=0.034455, w shape=(118,)
lambda=1.0: loss=0.036291, w shape=(118,)


In [60]:
# choose different lambdas to test
lambdas = np.logspace(-5, 2, 8)

print("Lambda values to test:")
print(lambdas)

Lambda values to test:
[1.e-05 1.e-04 1.e-03 1.e-02 1.e-01 1.e+00 1.e+01 1.e+02]


In [61]:
# use same threshold grid as least squares to compare model
thresholds = np.linspace(0.05, 0.5, 46)

print("\nThresholds:")
print(thresholds)


Thresholds:
[0.05 0.06 0.07 0.08 0.09 0.1  0.11 0.12 0.13 0.14 0.15 0.16 0.17 0.18
 0.19 0.2  0.21 0.22 0.23 0.24 0.25 0.26 0.27 0.28 0.29 0.3  0.31 0.32
 0.33 0.34 0.35 0.36 0.37 0.38 0.39 0.4  0.41 0.42 0.43 0.44 0.45 0.46
 0.47 0.48 0.49 0.5 ]


In [62]:
# define functions I will need for ridge regression later on
def train_ridge(x_tr, y_tr, fold, lambda_):
    """
    Train ridge regression on one training fold.

    Parameters
    x_tr : training feature matrix
    y_tr : training labels in {0, 1}
    fold : fold number
    lambda_ : ridge regularization strength

    Returns
    w : fitted ridge regression weights
    """

    w, _ = ridge_regression(
        y_tr,
        x_tr,
        lambda_
    )

    return w

def score_ridge(w, x_val):
    """
    Generate raw ridge regression scores.

    """

    scores = x_val @ w

    return scores

In [63]:
# cross validation for every lambda!
ridge_results = []

for lambda_ in lambdas:

    print("\n" + "=" * 60)
    print("Testing lambda =", lambda_)
    print("=" * 60)

    # Create a lambda-specific training function for cross_validate() to know which lambda to use
    def train_ridge_lambda(x_tr, y_tr, fold, lambda_=lambda_):
        return train_ridge(
            x_tr,
            y_tr,
            fold,
            lambda_
        )

    # cross-validation
    best_t, mean, std = cross_validate(
        train_fn=train_ridge_lambda,
        score_fn=score_ridge,
        x=tx,
        y01=y01,
        thresholds=thresholds,
        k=5,
        seed=42
    )

    # Save results - do I need this?
    ridge_results.append({
        "lambda": lambda_,
        "threshold": best_t,
        "precision": mean[0],
        "recall": mean[1],
        "f1": mean[2],
        "precision_std": std[0],
        "recall_std": std[1],
        "f1_std": std[2]
    })


Testing lambda = 1e-05
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.21 | F1 0.4105 ± 0.0039 | P 0.3321 ± 0.0031 | R 0.5373 ± 0.0058

Testing lambda = 0.0001
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.21 | F1 0.4105 ± 0.0039 | P 0.3322 ± 0.0031 | R 0.5373 ± 0.0057

Testing lambda = 0.001
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.21 | F1 0.4105 ± 0.0040 | P 0.3324 ± 0.0031 | R 0.5366 ± 0.0059

Testing lambda = 0.01
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.20 | F1 0.4102 ± 0.0031 | P 0.3231 ± 0.0026 | R 0.5618 ± 0.0045

Testing lambda = 0.1
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.20 | F1 0.4102 ± 0.0045 | P 0.3446 ± 0.0040 | R 0.5066 ± 0.0054

Testing lambda = 1.0
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.12 | F1 0.3986 ± 0.0033 | P 0.3212 ± 0.0027 | R 0.5253 ± 0.0046

Testing lambd

In [64]:
# find the best lambda based on CV F1
best_result = max(
    ridge_results,
    key=lambda result: result["f1"]
)

best_lambda = best_result["lambda"]
best_t_ridge = best_result["threshold"]

print("\n" + "=" * 60)
print("BEST RIDGE MODEL")
print("=" * 60)

print("Best lambda:", best_lambda)
print("Best threshold:", best_t_ridge)

print(
    "CV Precision:",
    best_result["precision"],
    "±",
    best_result["precision_std"]
)

print(
    "CV Recall:",
    best_result["recall"],
    "±",
    best_result["recall_std"]
)

print(
    "CV F1:",
    best_result["f1"],
    "±",
    best_result["f1_std"]
)


BEST RIDGE MODEL
Best lambda: 0.0001
Best threshold: 0.21000000000000002
CV Precision: 0.3321669003599603 ± 0.0031203566721645488
CV Recall: 0.537290767903365 ± 0.005749829871371092
CV F1: 0.41052974951657717 ± 0.003940360162018162


In [66]:
# display lambdas for sanity check
print("\nAll Ridge results:")

for result in ridge_results:
    print(
        f"lambda = {result['lambda']:.5g} | "
        f"threshold = {result['threshold']:.2f} | "
        f"F1 = {result['f1']:.4f} ± {result['f1_std']:.4f} | "
        f"P = {result['precision']:.4f} | "
        f"R = {result['recall']:.4f}"
    )


All Ridge results:
lambda = 1e-05 | threshold = 0.21 | F1 = 0.4105 ± 0.0039 | P = 0.3321 | R = 0.5373
lambda = 0.0001 | threshold = 0.21 | F1 = 0.4105 ± 0.0039 | P = 0.3322 | R = 0.5373
lambda = 0.001 | threshold = 0.21 | F1 = 0.4105 ± 0.0040 | P = 0.3324 | R = 0.5366
lambda = 0.01 | threshold = 0.20 | F1 = 0.4102 ± 0.0031 | P = 0.3231 | R = 0.5618
lambda = 0.1 | threshold = 0.20 | F1 = 0.4102 ± 0.0045 | P = 0.3446 | R = 0.5066
lambda = 1 | threshold = 0.12 | F1 = 0.3986 ± 0.0033 | P = 0.3212 | R = 0.5253
lambda = 10 | threshold = 0.05 | F1 = 0.3400 ± 0.0038 | P = 0.3557 | R = 0.3257
lambda = 100 | threshold = 0.05 | F1 = 0.0000 ± 0.0000 | P = 0.0000 | R = 0.0000


In [67]:
# train ridge regression model on ALL training data
w_ridge_final, loss_ridge_final = ridge_regression(
    y01,
    tx,
    best_lambda
)

print("\n" + "=" * 60)
print("FINAL RIDGE MODEL")
print("=" * 60)

print("Lambda:", best_lambda)
print("Loss:", loss_ridge_final)
print("w shape:", w_ridge_final.shape)

# calculate raw scores on test data
scores_test_ridge = tx_test @ w_ridge_final

print("\nTest scores shape:", scores_test_ridge.shape)
print("Minimum score:", scores_test_ridge.min())
print("Maximum score:", scores_test_ridge.max())
print("First 10 scores:", scores_test_ridge[:10])



FINAL RIDGE MODEL
Lambda: 0.0001
Loss: 0.033591448053551055
w shape: (118,)

Test scores shape: (109379,)
Minimum score: -0.13117608095767835
Maximum score: 0.7415615919002878
First 10 scores: [ 0.13506723 -0.07315871  0.07384135  0.06153393  0.05970748 -0.04212272
  0.02942058 -0.02292922  0.20848682  0.19338866]


In [68]:
# convert raw scores to 0/1
y_pred_test_ridge01 = (
    scores_test_ridge > best_t_ridge
).astype(int)

print("\nFirst 20 predictions:", y_pred_test_ridge01[:20])

print(
    "Test positive rate:",
    np.mean(y_pred_test_ridge01 == 1)
)

print(
    "Test negative rate:",
    np.mean(y_pred_test_ridge01 == 0)
)

print(
    "Number of positive predictions:",
    np.sum(y_pred_test_ridge01 == 1)
)

print(
    "Number of negative predictions:",
    np.sum(y_pred_test_ridge01 == 0)
)



First 20 predictions: [0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 1]
Test positive rate: 0.14296162883185987
Test negative rate: 0.8570383711681402
Number of positive predictions: 15637
Number of negative predictions: 93742


In [69]:
# convert 0/1 back to -1/+1
y_pred_test_ridge = 2 * y_pred_test_ridge01 - 1

print(
    "\nFirst 20 submission predictions:",
    y_pred_test_ridge[:20]
)

print(
    "Unique labels:",
    np.unique(y_pred_test_ridge)
)



First 20 submission predictions: [-1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1 -1  1]
Unique labels: [-1  1]


# Do I need cell below for submission file? #

sub = np.genfromtxt(
    "dataset/sample_submission.csv",
    delimiter=",",
    skip_header=1
)

test_ids = sub[:, 0].astype(int)

print("\nNumber of test IDs:", len(test_ids))
print("Number of predictions:", len(y_pred_test_ridge))


# Create submissions directory if it doesn't exist
import os

os.makedirs("submissions", exist_ok=True)


# Save Ridge submission
np.savetxt(
    "submissions/submission_ridge.csv",
    np.c_[test_ids, y_pred_test_ridge],
    delimiter=",",
    fmt="%d",
    header="Id,Prediction",
    comments=""
)

print("Ridge submission saved!")


# ------------------------------------------------------------
# 14. Verify submission
# ------------------------------------------------------------

print(
    np.genfromtxt(
        "submissions/submission_ridge.csv",
        delimiter=",",
        skip_header=1
    )[:10]
)

# 3.6 Linear SVM #

In [70]:
# had help with this! because we haven't learned it in class
def linear_svm(
    x,
    y,
    lambda_=1e-4,
    learning_rate=1e-2,
    epochs=10,
    batch_size=256,
    class_weight=True,
    seed=42
):
    """
    Train a linear SVM using mini-batch SGD.

    Parameters
    x : array-like, shape (N, D)
        Standardized training features.
    y : array-like, shape (N,)
        Binary labels in {0, 1}.
    lambda_ : float
        L2 regularization strength.
    learning_rate : float
        SGD learning rate.
    epochs : int
        Number of passes through the training data.
    batch_size : int
        Number of observations per mini-batch.
    class_weight : bool
        If True, give more weight to the minority class.
    seed : int
        Random seed.

    Returns
    w : array, shape (D,)
        Learned weights.

    b : float
        Learned intercept.
    """

    rng = np.random.default_rng(seed)

    x = np.asarray(x, dtype=float)
    y = np.asarray(y)

    N, D = x.shape

    # convert labels from 0,1 to -1, +1
    y_svm = 2 * y - 1

    #initialize weights and intercept
    w = np.zeros(D)
    b = 0.0
    
    # calculate class weights
    if class_weight:

        n_negative = np.sum(y == 0)
        n_positive = np.sum(y == 1)

        weight_negative = N / (2 * n_negative)
        weight_positive = N / (2 * n_positive)

    else:

        weight_negative = 1.0
        weight_positive = 1.0

    # do mini-batch SGD
    for epoch in range(epochs):

        # Shuffle the training observations
        indices = rng.permutation(N)

        x_shuffled = x[indices]
        y_shuffled = y_svm[indices]

        # Loop through mini-batches
        for start in range(0, N, batch_size):

            end = min(start + batch_size, N)

            x_batch = x_shuffled[start:end]
            y_batch = y_shuffled[start:end]
        
            # calculate SVM scores
            scores = x_batch @ w + b

            # calculate margins
            margins = y_batch * scores

            # find observations violating margings
            violating = margins < 1

            if np.any(violating):

                x_violate = x_batch[violating]
                y_violate = y_batch[violating]

                # Give the minority class more weight
                weights = np.where(
                    y_violate == 1,
                    weight_positive,
                    weight_negative
                )

                # gradient of weighted hinge loss
                gradient_w_hinge = -np.mean(
                    (weights * y_violate)[:, None] * x_violate,
                    axis=0
                )

                gradient_b_hinge = -np.mean(
                    weights * y_violate
                )

            else:

                gradient_w_hinge = np.zeros(D)
                gradient_b_hinge = 0.0

            # add L2 regularization
            gradient_w = (
                gradient_w_hinge
                + lambda_ * w
            )

            gradient_b = gradient_b_hinge

            # update parameters 
            w -= learning_rate * gradient_w
            b -= learning_rate * gradient_b

    return w, b

In [71]:
def score_svm(model, x_val):
    """
    Calculate probability-like scores for validation data.
    """

    w, b = model

    # Linear SVM decision score
    decision_scores = x_val @ w + b

    # Numerical stability for sigmoid
    decision_scores = np.clip(
        decision_scores,
        -50,
        50
    )

    # Convert scores to values between 0 and 1
    probabilities = 1 / (
        1 + np.exp(-decision_scores)
    )

    return probabilities

In [72]:
def train_svm(x_tr, y_tr, fold):

    return linear_svm(
        x=x_tr,
        y=y_tr,
        lambda_=1e-4,
        learning_rate=1e-2,
        epochs=10,
        batch_size=256,
        class_weight=True,
        seed=42 + fold
    )

In [73]:
# check that model runs first
best_t_svm, mean_svm, std_svm = cross_validate(
    train_fn=train_svm,
    score_fn=score_svm,
    x=x_train_final,
    y01=y01,
    thresholds=thresholds,
    k=5,
    seed=42
)

fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.50 | F1 0.3631 ± 0.0046 | P 0.2334 ± 0.0044 | R 0.8170 ± 0.0094


In [74]:
# test different lambda values
lambdas_svm = np.logspace(-5, -1, 5)

print("Lambda values to test:")
print(lambdas_svm)

Lambda values to test:
[1.e-05 1.e-04 1.e-03 1.e-02 1.e-01]


In [75]:
# cross validate each lambda
svm_results = []

for lambda_ in lambdas_svm:

    def train_svm_lambda(
        x_tr,
        y_tr,
        fold,
        lambda_=lambda_
    ):

        return linear_svm(
            x=x_tr,
            y=y_tr,
            lambda_=lambda_,
            learning_rate=1e-2,
            epochs=10,
            batch_size=256,
            class_weight=True,
            seed=42 + fold
        )

    best_t, mean, std = cross_validate(
        train_fn=train_svm_lambda,
        score_fn=score_svm,
        x=x_train_final,
        y01=y01,
        thresholds=thresholds,
        k=5,
        seed=42
    )

    svm_results.append({
        "lambda": lambda_,
        "threshold": best_t,
        "precision": mean[0],
        "recall": mean[1],
        "f1": mean[2],
        "precision_std": std[0],
        "recall_std": std[1],
        "f1_std": std[2]
    })

    print(
        f"lambda={lambda_:.5g} | "
        f"threshold={best_t:.3f} | "
        f"F1={mean[2]:.4f} ± {std[2]:.4f}"
    )

fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.50 | F1 0.3629 ± 0.0045 | P 0.2333 ± 0.0042 | R 0.8168 ± 0.0088
lambda=1e-05 | threshold=0.500 | F1=0.3629 ± 0.0045
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.50 | F1 0.3631 ± 0.0046 | P 0.2334 ± 0.0044 | R 0.8170 ± 0.0094
lambda=0.0001 | threshold=0.500 | F1=0.3631 ± 0.0046
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.50 | F1 0.3633 ± 0.0047 | P 0.2337 ± 0.0045 | R 0.8169 ± 0.0096
lambda=0.001 | threshold=0.500 | F1=0.3633 ± 0.0047
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.50 | F1 0.3632 ± 0.0046 | P 0.2336 ± 0.0043 | R 0.8164 ± 0.0093
lambda=0.01 | threshold=0.500 | F1=0.3632 ± 0.0046
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.50 | F1 0.3622 ± 0.0044 | P 0.2328 ± 0.0042 | R 0.8156 ± 0.0090
lambda=0.1 | threshold=0.500 | F1=0.3622 ± 0.0044


In [76]:
best_svm_result = max(
    svm_results,
    key=lambda result: result["f1"]
)

best_lambda_svm = best_svm_result["lambda"]
best_t_svm = best_svm_result["threshold"]

print("\nBest Linear SVM:")
print("Lambda:", best_lambda_svm)
print("Threshold:", best_t_svm)
print("Precision:", best_svm_result["precision"])
print("Recall:", best_svm_result["recall"])
print("F1:", best_svm_result["f1"])
print("F1 std:", best_svm_result["f1_std"])


Best Linear SVM:
Lambda: 0.001
Threshold: 0.5
Precision: 0.23367156987882418
Recall: 0.816876617773943
F1: 0.36333817576659866
F1 std: 0.004673800182022259


In [77]:
svm_final = linear_svm(
    x=x_train_final,
    y=y01,
    lambda_=best_lambda_svm,
    learning_rate=1e-2,
    epochs=10,
    batch_size=256,
    class_weight=True,
    seed=42
)

In [78]:
scores_test_svm = score_svm(
    svm_final,
    x_test_final
)

In [79]:
print("Test score shape:", scores_test_svm.shape)
print("Minimum score:", np.min(scores_test_svm))
print("Maximum score:", np.max(scores_test_svm))
print("First 10 scores:", scores_test_svm[:10])

Test score shape: (109379,)
Minimum score: 0.016986088804536233
Maximum score: 0.9902904346126428
First 10 scores: [0.54405708 0.03268291 0.30130683 0.39217043 0.29539915 0.06191521
 0.17641267 0.15855225 0.72444981 0.69945451]


In [80]:
y_pred_test_svm01 = (
    scores_test_svm > best_t_svm
).astype(int)

In [81]:
y_pred_test_svm = (
    2 * y_pred_test_svm01 - 1
)

In [82]:
print("First 20 predictions:")
print(y_pred_test_svm[:20])

print("\nUnique labels:")
print(np.unique(y_pred_test_svm))

print("\nPositive prediction rate:")
print(np.mean(y_pred_test_svm01))

print("\nNumber of positive predictions:")
print(np.sum(y_pred_test_svm01))

print("\nNumber of negative predictions:")
print(np.sum(y_pred_test_svm01 == 0))

First 20 predictions:
[ 1 -1 -1 -1 -1 -1 -1 -1  1  1  1 -1 -1 -1  1 -1  1 -1 -1  1]

Unique labels:
[-1  1]

Positive prediction rate:
0.3093189734775414

Number of positive predictions:
33833

Number of negative predictions:
75546


# do I need this again? to save for submission

submission_svm = np.column_stack(
    (
        np.arange(len(y_pred_test_svm)),
        y_pred_test_svm
    )
)

np.savetxt(
    "submissions/submission_linear_svm.csv",
    submission_svm,
    delimiter=",",
    header="Id,Prediction",
    comments="",
    fmt="%d"
)

print("Linear SVM submission saved.")

## 3.7 Logistic regression using gradient descent (y ∈ {0, 1})

In [34]:
# Step 1: choose gamma and max_iters

# The logistic loss is (L/4)-smooth, where L = largest eigenvalue of X^T X / N,
# so GD is stable for gamma <= 1 / (L/4).
L = np.linalg.eigvalsh(tx.T @ tx / N).max()
gamma_log = 1.0 / (L / 4)
print("gamma_log =", gamma_log)

# Training loss changes with # of iterations:
for iters in [100, 300, 1000, 3000]:
    w, loss = logistic_regression(y01, tx, np.zeros(D), iters, gamma_log)
    print(iters, "iters -> loss:", loss)
# max_iters = 1000 chosen, smallest after which the loss barely improves

gamma_log = 0.44824177578135965
100 iters -> loss: 0.22652056716242602
300 iters -> loss: 0.22319984714386884
1000 iters -> loss: 0.22274639884408168
3000 iters -> loss: 0.22273298969858896


<div style = "white-space: pre-line">Performance:
gamma_log = 0.44824177578135965
100 iters -> loss: 0.22652056716242602
300 iters -> loss: 0.22319984714386884
1000 iters -> loss: 0.22274639884408168
3000 iters -> loss: 0.22273298969858896 </div>

In [35]:
# Functions needed for Step 2
def train_logreg(x, y, fold):
    """Train logistic regression with GD on the training rows, return the weights w."""
    w, _ = logistic_regression(y, x, np.zeros(x.shape[1]), max_iters_log, gamma_log)
    return w

def score_logistic(w, x):
    """Predicted probability of heart attack for each row."""
    return sigmoid(x @ w)

In [36]:
# Step 2: 5-fold cross validation with the chosen parameters
thr_log = np.linspace(0.05, 0.6, 56)      # scores are probabilities in [0, 1]
max_iters_log = 1000                       # from Step 1

t_log, m_log, s_log = cross_validate(train_logreg,score_logistic,tx, y01, thr_log)
# if t_log is 0.05 or 0.6 (edge of the grid), widen the grid and rerun

fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.19 | F1 0.4234 ± 0.0039 | P 0.3433 ± 0.0028 | R 0.5522 ± 0.0076


<div style = "white-space: pre-line">Performance:
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.19 | F1 0.4234 ± 0.0039 | P 0.3433 ± 0.0028 | R 0.5522 ± 0.0076 </div>

In [46]:
# Step 3: final model on ALL training rows, then test predictions
w_log, loss_log = logistic_regression(y01, tx, np.zeros(D), 1000, gamma_log)
print("final train loss:", loss_log)

y_pred_log = np.where(sigmoid(tx_test @ w_log) > t_log, 1, -1)   # back to {-1, 1}

# predicted positive rate should be similar on train and test
print("predicted positive rate, test: ", np.mean(y_pred_log == 1))
print("predicted positive rate, train:", np.mean(sigmoid(tx @ w_log) > t_log))

os.makedirs("submissions", exist_ok=True)
create_csv_submission(test_ids, y_pred_log, "submissions/submission_logreg.csv")

KeyboardInterrupt: 

<div style = "white-space: pre-line">Performance:
final train loss (without penalty): 0.22274639884408168
predicted positive rate, test:  0.1423673648506569
predicted positive rate, train: 0.14209090770566993 </div>

## Regularized logistic regression using gradient descent (y ∈ {0, 1}, with regularization term λ∥w∥2)

In [38]:
# Functions needed for Step 1
def train_reglogreg(x, y, _fold):
    """Train regularized logistic regression with GD on the training rows, return the weights w.
    Uses the current values of lambda_log and gamma_reglog.
    _fold is not used (GD is deterministic), it is only there to match cross_validate."""
    w, _ = reg_logistic_regression(y, x, lambda_log, np.zeros(x.shape[1]), max_iters_log, gamma_reglog)
    return w

In [39]:
# Step 1: choose lambda with 5-fold CV (same folds as all other models)
# gamma and max_iters are reused from the logistic regression section.
# The penalty lambda * ||w||^2 adds 2*lambda to the curvature, so the safe step size
# becomes gamma = 1 / (L/4 + 2*lambda).
thr_log = np.linspace(0.05, 0.6, 56)      # scores are probabilities in [0, 1]

# Grid search: one full 5-fold CV per lambda
results_reglog = {}
for lambda_log in [1e-6, 1e-5, 1e-4, 1e-3, 1e-2]:
    gamma_reglog = 1.0 / (L / 4 + 2 * lambda_log)
    print("lambda", lambda_log)
    results_reglog[lambda_log] = cross_validate(train_reglogreg, score_logistic, tx, y01, thr_log)

# Pick the lambda with the highest mean F1
best_lambda = max(results_reglog, key=lambda l: results_reglog[l][1][2])
print("best lambda:", best_lambda)
print(f"reference, no regularization: F1 {m_log[2]:.4f} ± {s_log[2]:.4f}")

lambda 1e-06
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.19 | F1 0.4234 ± 0.0039 | P 0.3433 ± 0.0028 | R 0.5522 ± 0.0076
lambda 1e-05
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.19 | F1 0.4233 ± 0.0039 | P 0.3431 ± 0.0027 | R 0.5524 ± 0.0076
lambda 0.0001
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.19 | F1 0.4231 ± 0.0035 | P 0.3424 ± 0.0026 | R 0.5537 ± 0.0072
lambda 0.001
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.21 | F1 0.4231 ± 0.0043 | P 0.3548 ± 0.0033 | R 0.5241 ± 0.0072
lambda 0.01
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.24 | F1 0.4192 ± 0.0035 | P 0.3409 ± 0.0031 | R 0.5440 ± 0.0052
best lambda: 1e-06
reference, no regularization: F1 0.4234 ± 0.0039


<div style = "white-space: pre-line"> Performance:
lambda 1e-06
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.19 | F1 0.4234 ± 0.0039 | P 0.3433 ± 0.0028 | R 0.5522 ± 0.0076
lambda 1e-05
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.19 | F1 0.4233 ± 0.0039 | P 0.3431 ± 0.0027 | R 0.5524 ± 0.0076
lambda 0.0001
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.19 | F1 0.4231 ± 0.0035 | P 0.3424 ± 0.0026 | R 0.5537 ± 0.0072
lambda 0.001
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.21 | F1 0.4231 ± 0.0043 | P 0.3548 ± 0.0033 | R 0.5241 ± 0.0072
lambda 0.01
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.24 | F1 0.4192 ± 0.0035 | P 0.3409 ± 0.0031 | R 0.5440 ± 0.0052
best lambda: 1e-06
reference, no regularization: F1 0.4234 ± 0.0039 </div>

In [40]:
 # Step 2: CV results of the best lambda (already computed in step 1)
lambda_log = best_lambda
gamma_reglog = 1.0 / (L / 4 + 2 * lambda_log)
t_reglog, m_reglog, s_reglog = results_reglog[best_lambda]
print(f"threshold {t_reglog:.2f} | F1 {m_reglog[2]:.4f} ± {s_reglog[2]:.4f} "
      f"| P {m_reglog[0]:.4f} | R {m_reglog[1]:.4f}")

threshold 0.19 | F1 0.4234 ± 0.0039 | P 0.3433 | R 0.5522


<div style = "white-space: pre-line"> Performance:
threshold 0.19 | F1 0.4234 ± 0.0039 | P 0.3433 | R 0.5522
</div>

In [41]:
# Step 3: final model on ALL training rows, then test predictions
w_reglog, loss_reglog = reg_logistic_regression(y01, tx, lambda_log, np.zeros(D), max_iters_log, gamma_reglog)
print("final train loss (without penalty):", loss_reglog)

y_pred_reglog = np.where(sigmoid(tx_test @ w_reglog) > t_reglog, 1, -1)   # back to {-1, 1}

# sanity check: predicted positive rate should be similar on train and test
print("predicted positive rate, test: ", np.mean(y_pred_reglog == 1))
print("predicted positive rate, train:", np.mean(sigmoid(tx @ w_reglog) > t_reglog))

os.makedirs("submissions", exist_ok=True)
create_csv_submission(test_ids, y_pred_reglog, "submissions/submission_reglogreg.csv")

final train loss (without penalty): 0.2227465284482468
predicted positive rate, test:  0.14241307746459558
predicted positive rate, train: 0.1420878601794993


<div style = "white-space: pre-line"> Performance:
final train loss (without penalty): 0.2227465284482468
predicted positive rate, test:  0.14241307746459558
predicted positive rate, train: 0.1420878601794993
</div>

## Random forest

In [4]:
# Step 1: runtime of one forest tree (decides how many trees we can afford)
import time
edges = make_edges(x_train_final)
B = bin_data(x_train_final, edges)
n_bins = max(len(e) for e in edges) + 1
B_test = bin_data(x_test_final, edges)

max_features_rf = int(np.sqrt(B.shape[1]))     # standard choice: sqrt(number of features), here ~10

t0 = time.time()
build_forest(B, y01, 1, n_bins, max_depth=12, min_leaf=50, max_features=max_features_rf, seed=0)
print("one tree:", round(time.time() - t0, 1), "s")

one tree: 1.6 s


Output:
one tree: 1.6 s

In [6]:
# Step 2: choose max_depth and n_trees on one fold (fold 0), like for the single tree
folds = stratified_kfold_indices(y01, 5, 42)
val_idx, tr_idx = folds[0], np.concatenate(folds[1:])
thr = np.linspace(0.05, 0.7, 66)

for depth in [8, 12, 16]:
    forest = build_forest(B[tr_idx], y01[tr_idx], 50, n_bins, depth, 50, max_features_rf, seed=0)
    preds = np.array([predict_tree(tree, B[val_idx]) for tree in forest])   # (n_trees, n_val)
    # F1 when using only the first 10, 25, 50 trees: shows if more trees still help
    for n_t in [10, 25, 50]:
        f1 = best_f1(preds[:n_t].mean(axis=0), y01[val_idx], thr)
        print("depth", depth, "trees", n_t, "-> val F1", round(f1, 4))

""" #longer run time: 20min
for depth in [8, 12, 16]:
    for leaf in [20, 50, 100]:
        forest = build_forest(B[tr_idx], y01[tr_idx], 50, n_bins, depth, 50, max_features_rf, seed=0)
        preds = np.array([predict_tree(tree, B[val_idx]) for tree in forest])   # (n_trees, n_val)
        # F1 when using only the first 10, 25, 50 trees: shows if more trees still help
        for n_t in [10, 25, 50]:
            f1 = best_f1(preds[:n_t].mean(axis=0), y01[val_idx], thr)
            print("depth", depth, "trees", n_t, "-> val F1", round(f1, 4))
"""

depth 8 trees 10 -> val F1 0.3987
depth 8 trees 25 -> val F1 0.405
depth 8 trees 50 -> val F1 0.4073
depth 12 trees 10 -> val F1 0.4088
depth 12 trees 25 -> val F1 0.4114
depth 12 trees 50 -> val F1 0.4137
depth 16 trees 10 -> val F1 0.4103
depth 16 trees 25 -> val F1 0.4156
depth 16 trees 50 -> val F1 0.4167


'\nfor depth in [8, 12, 16]:\n    for leaf in [20, 50, 100]:\n        forest = build_forest(B[tr_idx], y01[tr_idx], 50, n_bins, depth, 50, max_features_rf, seed=0)\n        preds = np.array([predict_tree(tree, B[val_idx]) for tree in forest])   # (n_trees, n_val)\n        # F1 when using only the first 10, 25, 50 trees: shows if more trees still help\n        for n_t in [10, 25, 50]:\n            f1 = best_f1(preds[:n_t].mean(axis=0), y01[val_idx], thr)\n            print("depth", depth, "trees", n_t, "-> val F1", round(f1, 4))\n'

Performance:
depth 8 trees 10 -> val F1 0.3987 <br>
depth 8 trees 25 -> val F1 0.405 <br>
depth 8 trees 50 -> val F1 0.4073 <br>
depth 12 trees 10 -> val F1 0.4088 <br>
depth 12 trees 25 -> val F1 0.4114<br>
depth 12 trees 50 -> val F1 0.4137 <br>
depth 16 trees 10 -> val F1 0.4103<br>
depth 16 trees 25 -> val F1 0.4156<br>
depth 16 trees 50 -> val F1 0.4167 <br>

In [7]:
# Functions needed for Step 3
def train_rf(Bx, y, fold):
    """Train a random forest on the binned training rows.
    The fold number is used in the seed, so that bootstrap and feature sampling are reproducible."""
    return build_forest(Bx, y, n_trees_rf, n_bins, max_depth_rf, min_leaf_rf, max_features_rf, seed=42 + fold)

def score_rf(forest, Bx):
    """Average leaf probability over all trees."""
    return predict_forest(forest, Bx)

In [8]:
# Step 3: 5-fold CV with the chosen settings
n_trees_rf = 50          # chosen in step 2
max_depth_rf = 12        # chosen in step 2
min_leaf_rf = 50

t_rf, m_rf, s_rf = cross_validate(train_rf, score_rf, B, y01, thr)

fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.19 | F1 0.4154 ± 0.0032 | P 0.3307 ± 0.0023 | R 0.5586 ± 0.0051


<div style = "white-space: pre-line"> Performance:
fold 0 done
fold 1 done
fold 2 done
fold 3 done
fold 4 done
best threshold 0.19 | F1 0.4154 ± 0.0032 | P 0.3307 ± 0.0023 | R 0.5586 ± 0.0051
</div>

In [9]:
# Step 4: final forest on ALL training rows, then test predictions
forest_final = build_forest(B, y01, n_trees_rf, n_bins, max_depth_rf, min_leaf_rf, max_features_rf, seed=0)

y_pred_rf = np.where(predict_forest(forest_final, B_test) > t_rf, 1, -1)   # B_test: binned with TRAIN edges

print("predicted positive rate, test: ", np.mean(y_pred_rf == 1))
print("predicted positive rate, train:", np.mean(predict_forest(forest_final, B) > t_rf))

os.makedirs("submissions", exist_ok=True)
create_csv_submission(test_ids, y_pred_rf, "submissions/submission_rf.csv")

predicted positive rate, test:  0.14925168450982365
predicted positive rate, train: 0.1487954652810581


<div style = "white-space: pre-line"> Performance:
predicted positive rate, test:  0.14925168450982365
predicted positive rate, train: 0.1487954652810581 </div>